components/xperiencify/app/xperiencify.app.ts

```diff
@@ -1,13 +1,170 @@
+          studentEmail,
+        });
+        return tags;
+      },
+    },
+  },
   methods: {
-    // this.$auth contains connected account data
-    authKeys() {
-      console.log(Object.keys(this.$auth));
+    async _makeRequest({
+      $ = this, method = "get", path, params, data, ...opts
+    }) {
+      try {
+        return await axios($, {
+          url: `https://api.xperiencify.io/api/public${path}`,
+          method,
+          params: {
+            api_key: this.$auth.api_key,
+            ...params,
+          },
+          data,
+          ...opts,
+        });
+      } catch (e) {
+        console.error(e.response.data);
+        throw e;
+      }
+    },
+    async listCourses() {
+      return this._makeRequest({
+        path: "/coach/courses",
+      });
+    },
+    async listAllStudents() {
+      const courses = await this.listCourses();
+      const promises = courses.map(async (course) => this.getStudentsForCourse({
+        courseId: course.id,
+      }));
+      return (await Promise.all(promises)).flat().sort();
+    },
+    async listStudentTags({
+      $, studentEmail,
+    }) {
+      return this._makeRequest({
+        $,
+        path: "/student/tag/list",
+        data: {
+          student_email: studentEmail,
+        },
+      });
+    },
+    async addTagsToStudent({
+      $, studentEmail, tags,
+    }) {
+      return this._makeRequest({
+        $,
+        path: "/student/tag/manager",
+        method: "post",
+        data: {
+          student_email: studentEmail,
+          tagname: tags,
+        },
+      });
+    },
+    async removeTagsFromStudent({
+      $, studentEmail, tags,
+    }) {
+      return this._makeRequest({
+        $,
+        path: "/student/tag/manager",
+        method: "delete",
+        data: {
+          student_email: studentEmail,
+          tagname: tags,
+        },
+      });
+    },
+    async removeStudentFromAllCourses({
+      $, studentEmail,
+    }) {
+      return this._makeRequest({
+        $,
+        path: "/student/course/remove/all",
+        method: "post", // yes, it's supposed to be post
+        data: {
+          student_email: studentEmail,
+        },
+      });
+    },
+    async getStudentsForCourse({ courseId }) {
+      const courses = await this.listCourses();
+      const [
+        course,
+      ] = courses.filter((course) => course.id === courseId) || [];
+
+      if (!course) {
+        throw new Error(`Course ${courseId} not found`);
+      }
+      return course.users;
+    },
+    async addStudentToCourse({
+      $, courseId, studentEmail, firstName, lastName, password,
+    }) {
+      return this._makeRequest({
+        $,
+        path: "/student/create",
+        method: "post",
+        data: {
+          course_id: courseId,
+          student_email: studentEmail,
+          first_name: firstName,
+          last_name: lastName,
+          password,
+        },
+      });
+    },
```

Comment:
One suggestion here. I always tend to leave the job of assigning `data` and `params` by using `...args` or `...opts` object. Just like this:
```suggestion
    async addStudentToCourse(opts = {}) {
      return this._makeRequest({
        path: "/student/create",
        method: "post",
        ...opts,
      });
    },
```
In that way I don't have to touch the app file in this case `xperiencify.app.ts` in order to make modifications in the future for only one action/source component that uses this function. So you will only modify your action/source like this

```javascript
     await this.xperiencify.addStudentToCourse({
        $,
        data: {
          course_id: courseId,
          student_email: studentEmail,
          first_name: firstName,
          last_name: lastName,
          password,
        },
      });
```

Reply:
I like it! Thanks!